### ISSUE TITLE
Go monotonic time and serialization

### ISSUE THREAD

**Issue by Kubuxu** [ipfs/notes#254]
Go 1.9 introduces monotonic time to Golang. It is awesome but with one catch, serialization always removes the monotonic component. I hoped that monotonic time would solve problems with time based replay protection in WireGuard protocol as time is used as sort of ever incrementing counter and is only compared against our local clock (we send time with handshake and it is being resent to us).

This would mean that the time would be compared against time in the same session, thus not breaking guarantees required for the monotonicity. Unfortunately as any component of monotonic time is stripped during serialization, this can't be done using `time` package.

**Comment by diasdavid** [comment]
@Kubuxu I can't tell if you are proposing a solution in your problem description. AFAIU we are not using WireGuard in go-ipfs, is this related to a possible integration?

**Comment by Kubuxu** [comment]
It is just a rough note as we were thinking about using WG in the past and the usage of time was one of the main issues. Me and @lgierth though Go 1.9 could resolve this issue but it does not.

**Comment by lgierth** [comment]
Yes definitely want to use Wireguard as the crypto channel for datagrams.
